src/model/id.rs

```diff
@@ -331,6 +347,61 @@ pub(crate) mod snowflake {
     use serde::de::{Error, Visitor};
     use serde::{Deserializer, Serializer};
 
+    /// impl from std ([T]::as_chunks)
+    const fn as_chunks<T, const N: usize>(slice: &[T]) -> (&[[T; N]], &[T]) {
+        assert!(N != 0, "chunk size must be non-zero");
+        let len = slice.len() / N;
+        let (multiple_of_n, remainder) = slice.split_at(len * N);
+        let new_len = multiple_of_n.len() / N;
+        // SAFETY: We cast a slice of `new_len * N` elements into
+        // a slice of `new_len` many `N` elements chunks.
+        (unsafe { std::slice::from_raw_parts(multiple_of_n.as_ptr().cast(), new_len) }, remainder)
+    }
+
+    fn simple_parse(mut n: u64, s: &[u8]) -> u64 {
+        for &b in s {
+            n = n * 10 + (b - b'0') as u64;
+        }
+        n
+    }
+
+    fn parse_eight(s: [u8; 8]) -> u64 {
+        // reinterpret as u64 ("92233721" => 92233721)
+        let n = u64::from_le_bytes(s);
+        // combine 4 pairs of single digits:
+        // split pieces into odd and even
+        //  1_7_3_2_ (le repr)
+        // _2_3_2_9
+        // then combine
+        // _21_37_23_92 (le repr, each byte as 2 digits)
+        let n = ((n & 0x0f000f000f000f00) >> 8) + ((n & 0x000f000f000f000f) * 10);
+        // combine 2 pairs of 2 digits:
+        // split again
+        // _21___23__
+        // ___37___92
+        // then combine
+        // __14|137__36|7 (le repr, pipes separating bytes)
+        let n = ((n & 0x00ff000000ff0000) >> 16) + ((n & 0x000000ff000000ff) * 100);
+        // combine pair of 4 digits
+        // split again
+        // __14|137____ (then moved to ______14|137, as u64:3721)
+        // ______36|07 (as u64: 9223)
+        // then combine
+        ((n & 0x0000ffff00000000) >> 32) + ((n & 0x000000000000ffff) * 10000)
+    }
+
+    // this parse is 4x faster than [`str::parse`], see <https://github.com/serenity-rs/serenity/pull/2677#issue-2060912973>
+    pub fn parse(x: &(impl AsRef<[u8]> + ?Sized)) -> Option<NonZeroU64> {
+        NonZeroU64::new(match as_chunks(x.as_ref()) {
+            // 16, ..4 (most flakes are length 18, so this is on top)
+            (&[a, b], rest) => simple_parse(parse_eight(a) * 100000000 + parse_eight(b), rest),
+            // 8, 4..8
+            (&[a], rest) => simple_parse(parse_eight(a), rest),
+            // largest number is 20 digits, if this is a problem you have bigger problems
+            (_, rest) => simple_parse(0, rest),
```

Comment:
> Otherwise, this is okay to merge.

Actually, could you fall back to `str::parse` here or handle the integer overflow yourself for this case? The test has a dummy snowflake that's too big to parse into a `u64` (or just change the snowflake to be smaller).

Reply:
the test has a digit of length 24?

Reply:
oh, `123123123123123123123`

Reply:
So, im looking at some tests:

https://github.com/serenity-rs/serenity/blob/c5e6b3d923043e0028b8ef3c2315becbafd0670d/src/utils/mod.rs#L130-L131

I think we have a problem.